### ISSUE TITLE
Requests with POST is not cached

### ISSUE THREAD

**Issue by AshwinTayson** [kuitos/axios-extensions#2]
POST requests are not cached. It is requested for every request. Is there some limitation or Do I need to configure it differently?

**Comment by kuitos** [comment]
closed

**Comment by Hongbo-Miao** [comment]
Hi @kuitos 

I think sometimes it makes sense to cache POST requests.

One example I can find is

```
POST https://graph.microsoft.com/api/beta/me/getMailTips

{
    "EmailAddresses": [
        "jack@example.com", 
        "rose@example.com"
    ],
    "MailTipsOptions": "automaticReplies, mailboxFullStatus"
}
```

It returns Jack and Rose's mailtip.

BTW, thanks for brining us this library!

**Comment by kuitos** [comment]
This api was served by graphql?

**Comment by Hongbo-Miao** [comment]
No, it is REST API. Sorry, forget the link. https://developer.microsoft.com/en-us/graph/docs/api-reference/beta/api/user_getmailtips

**Comment by Hongbo-Miao** [comment]
@kuitos got it, thank you for the info!

**Comment by brendanmatkin** [comment]
For the sake of discussion: 

Technically the HTTP/1.1 RFC document says that POST is a cacheable method: https://tools.ietf.org/html/rfc7231#section-4.2.3.
MDN also says it is cacheable if `freshness information is included`: https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/POST

In my case I need to send data to be analyzed (as a big JSON) to the server in order to get a response (which can take several minutes to process) - I'd like to cache that response as long as the request body is the same. I don't mind having to manually forceUpdate, but it would be super nice to be *able* to cache the response to a POST.  
It's also possible that this pattern *shouldn't* be a RESTful interface.. 

I guess it might be a big change though, since it looks like this library is really structured around GETs and the buildURL helper - how would we handle the POST body if we were to update this project, right? 

Anyway I'd like to help but it's a bit over my head. If anyone can give me a few pointers, I can work on it and try to contribute. Thoughts?

**Comment by kuitos** [comment]
@brendanmatkin 
Thanks for your valuable information!
As more than one person had asked for the post cacheable feature, we may need to provide a mechanism to make user can define the cacheable logic by himself.
I'm going to think about how to do that in the near feature, and you can make any suggestion at any time.

**Comment by brendanmatkin** [comment]
@kuitos ok thanks! I'll chew on it.

For now I'm thinking (since it seems to be needed in only a few edge cases), that checking if the body has changed happens in the implementation (maybe using force-update), and axios-extensions just allows us to use POST as a method. (I guess this is probably what you mean by "define the caceable logic by himself").

This would probably be way easier to build but might lead to some confusion when things are automatically caching even though something has changed... maybe we add another option to flag it as manual caching logic?

**Comment by abhipanda** [comment]
@kuitos I feel a generic library like "axios-extensions" should cater to **all** the methods that axios supports without introducing any subjective presumptions about any methods. It should be on the user to determine on which rule of theory they follow. It should let the user choose ingredients of the hash e.g.  query, params, body, headers etc. You could just have "cache":true and additional hashing inputs from user.

Doing that would give this library a more generic, free from any assumptions.

**Comment by lkounadis** [comment]
@kuitos have a look a cachios, it has a nice implementation of it. mostly due to being able to cache the post request/response with a custom key you want
Good work though!

**Comment by roceroo** [comment]
> POST requests are not supported, and I think to fetch data with POST request was not a good idea...

I've created an package at npm to cache post and get request, just look for axios-post-cache

**Comment by gongjianh** [comment]
> @brendanmatkin 感谢您提供宝贵的信息！ 由于不止一个人要求后缓存功能，我们可能需要提供一种机制，让用户可以自己定义可缓存逻辑。 我会在near功能中考虑如何做到这一点，您可以随时提出任何建议。

Excuse me, has this function been added？

**Comment by gongjianh** [comment]
> @brendanmatkin Thanks for your valuable information! As more than one person had asked for the post cacheable feature, we may need to provide a mechanism to make user can define the cacheable logic by himself. I'm going to think about how to do that in the near feature, and you can make any suggestion at any time.

Excuse me, has this function been added？

**Comment by FloaJa** [comment]
我看到源码里已经有了一个支持POST缓存的方法, 但是并没有对外暴漏, 是遇到了什么无法解决的问题吗?
